src/Xrefcheck/System.hs

```diff
@@ -36,21 +42,67 @@ readingSystem = unsafePerformIO
 -- Check the respective env variable which is usually set in all CIs.
 askWithinCI :: IO Bool
 askWithinCI = lookupEnv "CI" <&> \case
-  Just "1"                       -> True
+  Just "1" -> True
   Just (map C.toLower -> "true") -> True
-  _                              -> False
+  _ -> False
+
+-- | A FilePath that has been canonicalized with 'System.Directory.canonicalizePath',
+-- and with indirections expanded as much as possible. Should be created via
+-- 'mkGlobPattern'.
+newtype CanonicalPath = UnsafeCanonicalPath
+  { unCanonicalPath :: FilePath
+  } deriving newtype (Show, Eq, Ord)
+
+canonicalizePath :: FilePath -> IO CanonicalPath
+canonicalizePath = (UnsafeCanonicalPath . expandIndirections <$>) . Directory.canonicalizePath
+  where
+  -- 'canonicalizePath' does not handle indirections when the file does not exist,
+  -- but we can expand as much as we can syntactically.
+  expandIndirections :: FilePath -> FilePath
+  expandIndirections = Posix.joinPath . reverse . expand 0 . reverse . Posix.splitDirectories
+
+  expand :: Int -> [FilePath] -> [FilePath]
+  expand acc (".." : xs) = expand (acc + 1) xs
+  expand acc ("." : xs) = expand acc xs
+  expand 0 (x : xs) = x : expand 0 xs
+  expand acc (_ : xs) = expand (acc - 1) xs
+  expand acc [] = replicate acc ".."
 
--- | Glob pattern relative to repository root. Should be created via @mkGlobPattern@
-newtype RelGlobPattern = RelGlobPattern FilePath
+takeDirectory :: CanonicalPath -> CanonicalPath
+takeDirectory (UnsafeCanonicalPath p) = UnsafeCanonicalPath $ Posix.takeDirectory p
+
+getDirsBetweenRootAndFile :: CanonicalPath -> CanonicalPath -> [CanonicalPath]
+getDirsBetweenRootAndFile root@(UnsafeCanonicalPath rootPath) file =
+  case getRelativeChild root (takeDirectory file) of
+    Just path -> UnsafeCanonicalPath
+      <$> scanl (Posix.</>) rootPath (Posix.splitDirectories path)
+    Nothing -> []
+
+getRelativeChild :: CanonicalPath -> CanonicalPath -> Maybe FilePath
+getRelativeChild (UnsafeCanonicalPath root) (UnsafeCanonicalPath child) =
+  dropWhile Posix.isPathSeparator <$> stripPrefix root child
+
+getRelativeOrAbsoluteChild :: CanonicalPath -> CanonicalPath -> FilePath
+getRelativeOrAbsoluteChild root child@(UnsafeCanonicalPath c) =
+  fromMaybe c (getRelativeChild root child)
+
+(</) :: CanonicalPath -> FilePath -> IO CanonicalPath
+UnsafeCanonicalPath p </ f = canonicalizePath $ p Posix.</> f
+infixr 5 </
```

Comment:
This operator is being used by our code with a Markdown file path link (POSIX syntax) on the right hand side. I was thinking about replacing here the path separator with something like

```haskell
map (\c -> 
  if isPathSeparator c
  then pathSeparator
  else c
)
```

because, in `FilePath.Windows`, both '\\' and '/' hold for `isPathSeparator`, but the `pathSeparator` value is '\\'.

Then I realised that the resulting path goes through `canonicalizePath`, which uses `normalise` that already does that.

Does this mean that the Windows problems that you mentioned are already solved? Do you have any problem in mind that could appear with this behaviour?

Reply:
In this case it is fine, yeah. Looks like all places where Markdown file path is used to produce `CanonicalPath` are safe simply because we use `canonicalizePath` so often to construct `CanonicalPath`s.

We have to be careful about the cases when we work with Markdown file path and produce another Markdown file path. How many such places do we have? If effectively none, if this `</` is the only place where we consume Markdown file path, then probably let's abandon my idea with newtype for such links and simply mention in haddock that the right hand side of `</` can accept filepath in any format.